### PR TITLE
Handling  deleting an uploading media in Aztec

### LINKED ISSUE #6456: Async media: upload not cancelled when media deleted in Aztec
### Expected behavior
When you're uploading a media item (video or image) in the Editor, and you delete it, the upload should be cancelled.

### Actual behavior
The upload continues and there's no way for you to cancel it.

### Steps to reproduce the behavior
1. start a new draft on Aztec
2. insert media item
3. delete the item
4. observe the "uploading media" notification is still going on, and if you exit the editor the posts list will show the progress bar as well

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #6456 

This PR takes care of 2 situations:
- when a user deletes one or more media items within Aztec, [now Aztec triggers an event](https://github.com/wordpress-mobile/AztecEditor-Android/pull/461) through the `OnMediaDeletedListener` interface that is being listened to by `AztecEditorFragment` within WPAndroid. This in turn calls the host of such fragment, `EditPostActivity`, which takes care of notifying the `UploadService` of such a thing, which in turn takes care of cancelling the upload as it's not needed anymore.

- this comes with one other situation that can come up from this, and it's the user can go back and `undo` this on Aztec, so you end up having the same media item again, but now the upload for such an item has been stopped (and wiped out from the UploadService altogether). The `EditorFragmentListener` now has a new method, `onUndoMediaCheck()`, where the implementation checks for those media items that are in the new Post content (_undoed_ content) marked with the `UPLOADING` class attribute but are not there in the `UploadService`. These are marked as `FAILED` so they can be retried by the user.

To test:
1. start a new draft
2. insert a media item - preferrably something big so you're able to test an in-progress upload cancelation.
3. while the media item is being uploaded, place the cursor to its right (or after it), and hit backspace to delete it
4. observe the upload is stopped
5. tap on the UNDO button
6. observe the media item appears again, now with the "failed"  overlay
7. Tap on it to start the upload again, as normal

Test this also with variants such as: exiting the editor in the middle of an upload (so the upoading notification is triggered) and coming back in again, or inserting several media items mixed with text, selecting them all and deleting it, check how the Posts list behaves, etc.
Maybe even test we haven't broken anything in the Visual Editor as well (as this should only work with Aztec)

cc @nbradbury

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -2328,6 +2340,45 @@ public void onMediaUploadCancelClicked(String localMediaId) {
     }
 
     @Override
+    public void onMediaDeleted(String localMediaId) {
+        if (!TextUtils.isEmpty(localMediaId)) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I don't understand why `localMediaId` is a string in several places when it should be an int, but as long as it's being passed around as a string we need to play it safe and catch any potential `NumberFormatException` if it fails to be parsed as an int. The simplest way to do that is use `StringUtils.stringToInt()`.

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Good observation @nbradbury  - that's because in several places we obtain the id from `wp-id` class attribute within the Post content, which is text.

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java [hunk 2]
```diff
@@ -652,39 +711,44 @@ public void onMediaUploaded(OnMediaUploaded event) {
...
-                                mPostUploadHandler.upload(updatedPost);
-                            }
-                        }
+                cleanUp();
+            }
+            stopServiceIfUploadsComplete();
+        }
+    }
+
+    private void cleanUp() {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: It would be good to have a comment describing the purpose of this method.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 3]
```diff
@@ -1548,6 +1559,30 @@ public static String resetUploadingMediaToFailed(Context context, @NonNull Strin
         return postContent;
     }
 
+    public static List<String> getMediaMarkedUploadingInPost(Context context, @NonNull String postContent) {
```
Comment (nbradbury) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: I'm being nitpicky, but this method seems to be misnamed since other methods that have `Post` is their name operate on a `PostModel`. How about `getMediaMarkedUploadingInPostContent`?

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java [hunk 4]
```diff
@@ -652,39 +711,44 @@ public void onMediaUploaded(OnMediaUploaded event) {
...
+            while (iterator.hasNext()) {
+                UploadingPost uploadingPost = iterator.next();
+                if (!UploadService.hasPendingOrInProgressMediaUploadsForPost(uploadingPost.postModel)) {
+                    if (uploadingPost.isCancelled) {
+                        // Finished all media uploads for a post upload that was cancelled (probably because
+                        // it was re-opened in the editor) - we can just remove it from the list now
+                        iterator.remove();
+                    } else {
+                        // Fetch latest version of the post, in case it has been modified elsewhere
+                        PostModel latestPost = mPostStore.getPostByLocalPostId(uploadingPost.postModel.getId());
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: Do we know that all changes to the post will already have been saved to the store when this is called?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: addressed this one in a765480, but would probably be good to open a different PR to only take care of this in particular (handling of mediaIds throughout the app)

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java (hunk 2 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: added in 24f0b42

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java (hunk 4 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: Another good observation - this piece of code actually was written for the case when we're uploading things in the background. For the case where the user is actually editing the post, the Editor always has the last say on this, so chances of losing data are minimum (i.e. they can always stop the app while in the editor and before the autosave expires, that would be the only window for data loss in this case, unless they insert some media in which case we save things again as per https://github.com/wordpress-mobile/WordPress-Android/pull/6499).

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java (hunk 3 above)
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: 👍 addressed in 44e9b69

**Comment by mzorz** [comment]
Closing this in favor of #6519
